src/gateways/discord_gateway.ts

```diff
@@ -0,0 +1,57 @@
+import { Client, GuildMember, Guild, User } from "discord.js";
+
+
+export class DiscordGateway {
+    private static instance: DiscordGateway;
+
+    private client: Client;
+
+    private constructor() {
+        this.client = new Client();
+        this.client.login(process.env.BOT_TOKEN);
+
+        this.client.on("error", (error: Error) => {
+            console.error("Catastrophic error occurred on Discord.js client", error);
+            console.error("Exiting");
+            process.exit(1);
+        });
+
+        this.client.on("disconnect", () => {
+            console.error("server disconnected...");
+        });
+    }
+
+    static get getDiscordClient() {
+        if (!DiscordGateway.instance) {
+            DiscordGateway.instance = new DiscordGateway();
+        }
+
+        return DiscordGateway.instance.client;
+    }
+
+    static get getInstance() {
+        if (!DiscordGateway.instance) {
+            DiscordGateway.instance = new DiscordGateway();
+        }
+
+        return DiscordGateway.instance;
+    }
+
+    setPresenceUpdateHandler(handler: (oldMember: GuildMember, newMember: GuildMember) => Promise<void>) {
+        this.client.on("presenceUpdate", handler);
+    }
+
+    getAllAvailableGuilds() {
+        console.info("Getting all available guilds");
+        return DiscordGateway.instance.client.guilds.map((value: Guild) => value);
+    }
+
+    getViableVoiceChannelForUser(user: User): GuildMember {
```

Comment:
Where is this used

Reply:
@pruales in `PresenceManagement`.

Reply:
what are guilds?

Reply:
also why do you have `!!` in the filter? What if they have multiple guilds with voice channels?

Reply:
Also so in `presenceManagement` you send in the user and then you return the user here which is then used in voice which checks if voice channel is available?

Reply:
> what are guilds?

servers that the bot has access to. pretty sure

Reply:
> also why do you have `!!` in the filter? What if they have multiple guilds with voice channels?

genius